src/codeu/chat/util/Tokenizer.java

```diff
@@ -0,0 +1,91 @@
+/*
+* This class parses a string to retrive tokens
+* and determines whether they are surrounded by qoutations or not
+*
+* author: Isaac Martins
+* version: May 19, 2017
+*/
+package codeu.chat.util;
+
+import java.io.IOException;
+
+public final class Tokenizer {
+  //member variables
+  private StringBuilder token;
+  private String source;
+  private int at;
+
+  //functions (methods)
+  public Tokenizer(String source) {
+    this.source = source;
+    at = 0;
+    token = new StringBuilder();
+  }
+
+/**
+*@return returns the next character in the token/string
+*/
+  public String next() throws IOException {
+      //Skip preceding whitespace
+      while (remaining() > 0 && Character.isWhitespace(peek())) {
+        //ignore result because it's whitespace
+        read();
+      }
+      //if there are no characters remaining return null
+      if (remaining() <= 0) {
+        return null;
+      } else if (peek() == '\"') {
+        //read a token surrounded by qoutes
+        return readWithQuotes();
+      } else {
+        //read a token surrounded with no qoutes
+        return readWithNoQuotes();
+      }
+  }
+
+  //Helper functions (methods)
+
+  //finds number of remaining characters in string
+  private int remaining() {
+    return (source.length() - at);
+  }
+
+  //looks/checks for character in source at index
+  private char peek() throws IOException {
+   if (at < source.length()) {
+      return source.charAt(at);
+    } else {
+      throw new IOException();
+    }
+  }
+
+  //reads tokens
+  private char read() throws IOException {
+    final char c = peek();
+    at =+ 1;
```

Comment:
I found the bug! It's on this line. The problem is that `at =+ 1` looks an awful lot like `at += 1`... but it's not.

If you're curious how I found it, I didn't use any high-tech debugging tools, just a boatload of print statements. Check out my little debugging branch: https://github.com/lysanderp41/CodeUSummer2017/compare/Isaac's-Branch...java_heap_space_error

Unfortunately, even after fixing that, it still looks to me like one of the unit tests is failing. Still I figured I should at least pass along this pointer to help get you unstuck.